### DISCUSSION TITLE
What's the difference between normal tests and integration tests?

### DISCUSSION THREAD

**Discussion by EvaristeGalois11** [mapstruct/mapstruct#3205] [General]
Hi,
i'm trying to understand a bit further the structure of the project.

I noticed that you have two sets of tests: the "normal" ones inside the `mapstruct-processor` module and the ones inside the `mapstruct-integrationtest` module.

What is exactly the difference between this two kind of tests? And more specifically when one should write an integration test instead of a normal one?

Thanks!

**Comment by filiphr** [comment]
The difference between the "normal" and integration tests is in the way that they are executed. The "normal" tests are in a purists mindset also integration test. A normal MapStruct `@ProcessorTest` is test which is invoking a compiler, which invokes the MapStruct annotation processor, that will generate the java code and we then test the functionality of the generated mapper. This allows us to more easily refactor the MapStruct processor without the need to change the tests. The `@ProcessorTest` also allows you to test negative paths e.g. compiler errors, warnings etc. We have [Testing MapStruct](https://mapstruct.org/development/testing-mapstruct/) in our documentation that explains this a bit as well. The text there is for our old mechanism (when we were using JUnit 4), but the concepts are the same.

An integration test (inside the `mapstruct-integrationtest` module) is a test which is testing in the same way as a user would use MapStruct. i.e. through Maven or Gradle. Those tests are there to test things like Lombok integration, some newer Java language specifics etc. We also have a test which is running all the non `Erroneous` mappers in the `mapstruct-processor` tests and validating that it works with different Java versions.

>And more specifically when one should write an integration test instead of a normal one?

In 99% of the time a normal test should be written. If you are unsure then feel free to explain what you want to test and we can help you decide

**Reply by EvaristeGalois11** [reply]
Thank you so much, this is very clear! I will open a draft PR for what I'm working on so you can help me decide what's best to do.
